File: android/core/src/main/kotlin/mesh/core/protocol/Protocol.kt
/**
 * Mesh Messenger Core — Wire Protocol
 *
 * Platform-agnostic packet format matching the core skill spec.
 * This is a pure Kotlin module with NO Android imports.
 */

package mesh.core.protocol

import kotlinx.serialization.Serializable
import kotlinx.serialization.protobuf.ProtoNumber

/** Packet types matching the wire format spec */
enum class PacketType(val value: Int) {
    DATA(0),
    ACK(1),
    HELLO(2),
    ROUTE_PROBE(3),
    PREKEY_REQUEST(4),
    PREKEY_BUNDLE(5),
    FRAGMENT(6);

    companion object {
        fun fromValue(value: Int): PacketType =
            entries.firstOrNull { it.value == value }
                ?: throw IllegalArgumentException("Unknown packet type: $value")
    }
}

/** Transport types */
enum class TransportType {
    BLE, WIFI_DIRECT, WIFI_AWARE, MPC, WEBRTC, RELAY
}

/** Battery class for routing decisions */
enum class BatteryClass(val priority: Int) {
    MAINS(5),
    HIGH(4),
    NORMAL(3),
    LOW(2),
    EPHEMERAL(1);
}

/** Message delivery status */
enum class MessageStatus {
    PENDING, SENT, DELIVERED, READ, FAILED
}

/** 16-byte Node ID = first 16 bytes of SHA-256(Ed25519 public key) */
@JvmInline
value class NodeId(val bytes: ByteArray) {
    init {
        require(bytes.size == 16) { "NodeId must be exactly 16 bytes, got ${bytes.size}" }
    }

    fun toHex(): String = bytes.joinToString("") { "%02x".format(it) }
    fun shortId(): String = toHex().take(8).uppercase()

    companion object {
        val BROADCAST = NodeId(ByteArray(16) { 0xFF.toByte() })
        val ZERO = NodeId(ByteArray(16))

        fun fromHex(hex: String): NodeId {
            require(hex.length == 32) { "NodeId hex must be 32 chars" }
            val bytes = hex.chunked(2).map { it.toInt(16).toByte() }.toByteArray()
            return NodeId(bytes)
        }
    }

    override fun toString(): String = shortId()
}

/** 8-byte random packet identifier for dedup */
@JvmInline
value class PacketId(val bytes: ByteArray) {
    init {
        require(bytes.size == 8) { "PacketId must be exactly 8 bytes" }
    }

    fun toHex(): String = bytes.joinToString("") { "%02x".format(it) }

    companion object {
        fun random(): PacketId {
            val bytes = ByteArray(8)
            java.security.SecureRandom().nextBytes(bytes)
            return PacketId(bytes)
        }
    }
}

/** Core mesh packet — matches the wire format spec */
data class MeshPacket(
    val version: UByte = PROTOCOL_VERSION,
    val type: PacketType,
    val packetId: PacketId,
    val src: NodeId,
    val dst: NodeId,
    val ttl: UByte,
    val hopCount: UByte = 0u,
    val timestamp: UInt,        // unix seconds
    val payload: ByteArray,     // ciphertext — routing nodes MUST NOT read
    val signature: ByteArray? = null,
) {
    companion object {
        /** Fixed header size: 1+1+8+16+16+1+1+4+2 = 50 bytes */
        const val HEADER_SIZE = 50
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is MeshPacket) return false
        return packetId.bytes.contentEquals(other.packetId.bytes)
    }

    override fun hashCode(): Int = packetId.bytes.contentHashCode()
}

/** Fragment header for reassembly */
data class FragmentHeader(
    val messageId: PacketId,  // groups fragments
    val index: UByte,         // 0-based
    val total: UByte,         // total count
)

/** HELLO beacon payload */
data class HelloPayload(
    val nodeId: NodeId,
    val batteryClass: BatteryClass,
    val transports: Set<TransportType>,
    val bloomFilter: ByteArray,
    val relayCapable: Boolean,
)

// ── Constants ──────────────────────────────────────────────────────────────

val PROTOCOL_VERSION: UByte = 1u
const val DEFAULT_TTL: Int = 7
const val MAX_PACKET_AGE_SECONDS: Long = 24 * 60 * 60  // 24 hours
const val BLE_FRAGMENT_SIZE: Int = 160
const val HELLO_INTERVAL_MS: Long = 20_000  // 20 seconds (between 15-30s)

File: ios/Sources/MeshCore/Protocol.swift
import Foundation

public enum PacketType: UInt8 {
    case data = 0
    case ack = 1
    case hello = 2
    case routeProbe = 3
    case prekeyRequest = 4
    case prekeyBundle = 5
    case fragment = 6
}

public enum TransportType: String, Codable {
    case ble = "ble"
    case wifiDirect = "wifi-direct"
    case wifiAware = "wifi-aware"
    case mpc = "mpc"
    case webrtc = "webrtc"
    case relay = "relay"
}

public enum BatteryClass: String, Codable {
    case mains = "mains"
    case high = "high"
    case normal = "normal"
    case low = "low"
    case ephemeral = "ephemeral"

    public var priority: Int {
        switch self {
        case .mains: return 5
        case .high: return 4
        case .normal: return 3
        case .low: return 2
        case .ephemeral: return 1
        }
    }
}

public enum MessageStatus: String, Codable {
    case pending = "pending"
    case sent = "sent"
    case delivered = "delivered"
    case read = "read"
    case failed = "failed"
}

public typealias NodeId = Data
public typealias PacketId = Data

public let PROTOCOL_VERSION: UInt8 = 1
public let DEFAULT_TTL: UInt8 = 7
public let MAX_PACKET_AGE_SECONDS: TimeInterval = 24 * 60 * 60
public let BLE_FRAGMENT_SIZE = 160
public let BROADCAST_DST = Data(repeating: 0xFF, count: 16)

public struct MeshPacket: Equatable {
    public let version: UInt8
    public let type: PacketType
    public let packetId: PacketId
    public let src: NodeId
    public let dst: NodeId
    public let ttl: UInt8
    public let hopCount: UInt8
    public let timestamp: UInt32
    public let payload: Data
    public let signature: Data?

    public static let headerSize = 50

    public init(
        version: UInt8 = PROTOCOL_VERSION,
        type: PacketType,
        packetId: PacketId,
        src: NodeId,
        dst: NodeId,
        ttl: UInt8 = DEFAULT_TTL,
        hopCount: UInt8 = 0,
        timestamp: UInt32,
        payload: Data,
        signature: Data? = nil
    ) {
        self.version = version
        self.type = type
        self.packetId = packetId
        self.src = src
        self.dst = dst
        self.ttl = ttl
        self.hopCount = hopCount
        self.timestamp = timestamp
        self.payload = payload
        self.signature = signature
    }

    public static func == (lhs: MeshPacket, rhs: MeshPacket) -> Bool {
        return lhs.packetId == rhs.packetId
    }
}

public struct FragmentHeader {
    public let messageId: PacketId
    public let index: UInt8
    public let total: UInt8
}

public struct HelloPayload: Codable {
    public let nodeId: NodeId
    public let batteryClass: BatteryClass
    public let transports: [TransportType]
    public let bloomFilter: Data
    public let relayCapable: Bool
}

extension Data {
    public func toHex() -> String {
        return map { String(format: "%02x", $0) }.joined()
    }

    public func shortId() -> String {
        return toHex().prefix(8).uppercased()
    }
}
